### PR TITLE
[12.0] [ADD] Account Clearance Plan

### PR DESCRIPTION
**Pull request by qgroulard**
Adds the opportunity to create clearance plans in order to reorganize and spread debts (own or customer ones) over a longer period of time.

clearance_plan

### CODE REVIEW

File: account_clearance_plan/models/account_invoice.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+class AccountInvoice(models.Model):
+
+    _inherit = "account.invoice"
+
+    def get_clearance_plan_wizard(self):
+        move_lines = self.env["account.move.line"]
+        for rec in self:
+            move_lines |= rec.move_id.line_ids.filtered(
+                lambda l: l.account_id == rec.account_id
+            )
```
Comment (sbidoul) [inline account_clearance_plan/models/account_invoice.py]: nit:

```suggestion
        move_lines = self.mapped("move_id.line_ids").filtered(
            lambda l: l.account_id == rec.account_id
        )
```

File: account_clearance_plan/readme/DESCRIPTION.rst [hunk 2]
```diff
@@ -0,0 +1,5 @@
+This module allows to reorganize customers debts in case of difficult financial conditions in order to spread the debts over a longer period of time.
+The module can also be used in a similar way for the user's company in case their suppliers offer them the opportunity to discuss a clearance plan.
+
+One can decide the debts to be part of the clearance plan (all partner's debts or some of them only) and split this amount into new financial commitments at new dates.
+Please notice that one clearance plan = one financial account.
```
Comment (sbidoul) [inline account_clearance_plan/readme/DESCRIPTION.rst]:
```suggestion
Please notice that one clearance plan = one payable/receivable account.
```

File: account_clearance_plan/readme/USAGE.rst [hunk 3]
```diff
@@ -0,0 +1,11 @@
+To use this module, you need to:
+
+#. Go to *Invoicing > Accounting > Journal Items* (or any invoices)
```
Comment (sbidoul) [inline account_clearance_plan/readme/USAGE.rst]:
```suggestion
#. Go to *Invoicing > Accounting > Journal Items*, or *Invoicing > Vendor Bills*, or *Invoicing > Customer Invoices*
```

File: account_clearance_plan/models/account_invoice.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+        context = {"move_line_ids": move_lines.ids}
```
Comment (sbidoul) [inline account_clearance_plan/models/account_invoice.py]:
```suggestion
        context = {"active_ids": move_lines.ids, "active_model": "accout.move.line"}
```

Would this work and simplify the `default_get` in the wizard?

File: account_clearance_plan/wizard/account_clearance_plan.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+                key2 = "credit"
+            else:
+                key1 = "credit"
+                key2 = "debit"
+            new_lines |= line.with_context(check_move_validity=False).copy(
+                default={
+                    "move_id": move.id,
+                    key1: 0,
+                    key2: abs(line.amount_residual),
+                }
```
Comment (sbidoul) [inline account_clearance_plan/wizard/account_clearance_plan.py:126]: Don't we need a specific move line `name` here?

File: account_clearance_plan/wizard/account_clearance_plan.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+                key1 = "debit"
+                key2 = "credit"
+            else:
+                key1 = "credit"
+                key2 = "debit"
+            new_lines |= line.with_context(check_move_validity=False).copy(
+                default={
+                    "move_id": move.id,
+                    key1: 0,
+                    key2: abs(line.amount_residual),
```
Comment (sbidoul) [inline account_clearance_plan/wizard/account_clearance_plan.py]:
```suggestion
                    "debit": abs(line.amount_residual) if line.credit > 0 else 0,
                    "credit": abs(line.amount_residual) if line.debit > 0 else 0,
```

File: account_clearance_plan/wizard/account_clearance_plan.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+        required=True,
+        help="Journal of the new entry.",
+    )
+    amount_to_allocate = fields.Float(string="Total Amount to Allocate", readonly=True)
+    amount_allocated = fields.Float(
+        string="Amount Allocated", compute="_compute_amount_allocated"
+    )
+    clearance_plan_line_ids = fields.One2many(
+        comodel_name="account.clearance.plan.line", inverse_name="clearance_plan_id"
+    )
```
Comment (sbidoul) [inline account_clearance_plan/wizard/account_clearance_plan.py:56]: Do we need to populate `ref`, `narration` in the generated move?

**Review by sbidoul — COMMENTED** [review]
Looks nice. A few minor comments.

We need to pay careful attention to the move and move line names that are generated so the partner ledger and credit control operations remain legible.

File: account_clearance_plan/wizard/account_clearance_plan.py (hunk 7 above)
Comment (qgroulard) [reply account_clearance_plan/wizard/account_clearance_plan.py:56]: It's quite difficult to give a ref that makes sense, the clearance plan may be created based on many invoices or move lines. Do you have any suggestion ?

I've implemented the rest of your comments.

Comment (sbidoul) [reply account_clearance_plan/wizard/account_clearance_plan.py:56]: Can the user provide it?

Comment (qgroulard) [reply account_clearance_plan/wizard/account_clearance_plan.py:56]: Indeed !
The user can now provide it in the wizard as well as the `narration`.

**Comment by gva-acsone** [comment]
Works really well ! 👍 

Just a small observation: it is currently impossible to open the 'clearance plan journal entry' starting from the invoice.
The 'view' button on the reconciled entry doesn't work. Would be really nice to make this work.

Sans titre



For the rest everything seems to work just fine (functional test) and this functionality will be really useful.


Possible improvements for the future:

1. Add a field 'Amount still to allocate' above the Clearance Plan Board (difference between amount_to_allocate and amount_allocated)
2. Add a 'description' column in the clearance plan table. Each clearance plan line can be associated to a description, which will be automatically added in the 'label' field on the related journal item. This new field can be automatically pre-filled with the "Default Clearance Plan Move Line Name" defined in the accounting settings.

**Review by sbidoul — APPROVED** [review]

**Comment by qgroulard** [comment]
> Works really well ! +1
> 
> Just a small observation: it is currently impossible to open the 'clearance plan journal entry' starting from the invoice.
> The 'view' button on the reconciled entry doesn't work. Would be really nice to make this work.
> 
> Sans titre
> 
> For the rest everything seems to work just fine (functional test) and this functionality will be really useful.
> 
> Possible improvements for the future:
> 
>     1. Add a field 'Amount still to allocate' above the Clearance Plan Board (difference between amount_to_allocate and amount_allocated)
> 
>     2. Add a 'description' column in the clearance plan table. Each clearance plan line can be associated to a description, which will be automatically added in the 'label' field on the related journal item. This new field can be automatically pre-filled with the "Default Clearance Plan Move Line Name" defined in the accounting settings.

@gva-acsone 
I've fixed the button and implemented your suggestions.
Thank you for the feedback !

**Review by sbejaoui — APPROVED** [review]

**Comment by qgroulard** [comment]
@OCA/core-maintainers , This PR is ready to merge

**Pull request merged**